### PR TITLE
Use resolved dataset options instead of dataset configuration

### PR DESCRIPTION
**Pull request by stockiNail**
This PR is using the resolved dataset options instead of dataset configuration in order to leverage on fallback to element options.

This also adds key, groups and treeLeafKey options to chart option for the same reason.

### TODO

 - [ ] Wait for #144

### CODE REVIEW

File: src/controller.js [hunk 1]
```diff
@@ -131,13 +131,15 @@ export default class TreemapController extends DatasetController {
     const meta = this.getMeta();
     const dataset = this.getDataset();
     const dpr = this.chart.currentDevicePixelRatio;
+    const options = this.options;
+    const datasetOptions = this.resolveDatasetElementOptions(mode);
```
Comment (kurkle) [inline src/controller.js:135]: What this essentially does now, is make the dataset options fall through treemap element options.
The required defaults should just be moved to the controller from element instead.

I hope I did not miss some other side effect.

**Comment by kurkle** [comment]
> EDIT: I went more in deep and I understood how datasets[type] is working :(

Hmm, not sure what you found out. Can you explain? :)

File: src/controller.js (hunk 1 above)
Comment (stockiNail) [reply src/controller.js:135]: > The required defaults should just be moved to the controller from element instead.

See test case of this PR: https://github.com/kurkle/chartjs-chart-treemap/pull/151/files#diff-9bf86eb940737e9d2f1b9a4743ca173f517e202cd3a1a4cfcb7c7072629fb901

If the user sets options at `elements.treemap` level in chart config, they are ignore, moving defaults in the `controller.defaults`.

Comment (stockiNail) [reply src/controller.js:135]: I have seen that `this.options` (controller/dataset options) has got the fallback to `datasets.treemap` but not to the `elements.treemap`. Instead, the dataset element options (`this.resolveDatasetElementOptions(mode);`) has the fallbacks to `datasets.treemap` and `elements.treemap` but it uses ONLY the keys of `elements.treemap`, ignoring the `datasets.treemap` keys.

**Comment by stockiNail** [comment]
@kurkle I had a look.

In the controller, `controller.options` is able to resolve going to `datasets.treemap`. 

> > EDIT: I went more in deep and I understood how datasets[type] is working :(
> 
> Hmm, not sure what you found out. Can you explain? :)

I have seen that this.options (controller/dataset options) has got the fallback to datasets.treemap but not to the elements.treemap. Instead, the dataset element options (this.resolveDatasetElementOptions(mode);) has the fallbacks to datasets.treemap and elements.treemap but it uses ONLY the keys of elements.treemap, ignoring the datasets.treemap keys.

**Comment by stockiNail** [comment]
Anyway I'm doing some other tests. Let me try again what you proposed

**Comment by stockiNail** [comment]
Test about what I wrote above (elements.treemap ignored).

```javascript
TreemapController.defaults = {
  dataElementType: 'treemap',
  captions: {
    display: true // <- DEFAULT
  },
  animations: {
    numbers: {
      type: 'number',
      properties: ['x', 'y', 'width', 'height']
    },
  },
};
```

Chart config:

```javascript
const config = {
  type: 'treemap',
  data: {
    datasets: [{
      ....
      captions: {
        align: 'center',
      },
      ....
    }]
  },
  options: {
    elements: {
	  treemap: {
	    captions: {
		  display: false, // <- CAPTION SHOULD HAVE display = false
		}
	  }
	}
  }
};
```

Inside the controller, `update(mode)` method:

```javascript
  update(mode) {
    const meta = this.getMeta();
    const dataset = this.getDataset();
    const dpr = this.chart.currentDevicePixelRatio;
    console.log(this.options.captions.display); // SHOULD EMIT false BUT IT EMITS true
```

**Comment by stockiNail** [comment]
The only options seems to move all configuration, leaving to the element only the following:

- borderRadius
- borderColor
- backgroundColor

The issue is `borderWidth` which is needed anyway to create data but I think it should be an option of the element.
Going to this, it's a breaking change, therefore we can postpone to version 3. I have tried do not create a breaking change.

**Comment by stockiNail** [comment]
I think the best way is to go to data parsing (leveraging on controller options) and move the size calculation when elements are created (using data element options, already calculated). in the controller options will have only the dataset keys (key, groups, treeLeafKey and additionalKeys from PR #144).
